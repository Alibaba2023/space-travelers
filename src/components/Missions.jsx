import React, { useEffect } from 'react';
import { Table, Button } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { AllMissions, fetchMissions } from '../redux/missions/missionSlice';
import '../style/missions.css';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector(AllMissions);

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  return (
    <Table striped bordered hover className="missions-table">
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th>Join/Leave</th>
        </tr>
      </thead>
      <tbody>
        {missions.map(({ mission_id, mission_name, description }) => (
          <tr key={mission_id}>
            <td>{mission_name}</td>
            <td className="description">{description}</td>
            <td>Not a Member / Active Member</td>
            <td>
              <Button variant="outline-success" type="button">Join</Button>
              <Button variant="outline-danger" type="button">Leave</Button>
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

export default Missions;
