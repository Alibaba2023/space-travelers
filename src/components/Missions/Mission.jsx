import React from 'react';
import { useDispatch } from 'react-redux';
import { Button, Badge } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { reserveMission } from '../../redux/missions/missionSlice';

const Mission = ({ mission }) => {
  const {
    id,
    name,
    description,
    reserved,
  } = mission;
  const dispatch = useDispatch();

  const handleButtonClick = () => {
    dispatch(reserveMission(id));
  };

  return (
    <>
      <td>{name}</td>
      <td className="description">{description}</td>
      <td>
        {reserved ? (
          <Badge bg="success">Active Member</Badge>
        ) : (
          <Badge bg="secondary">Not A Member</Badge>
        )}
      </td>
      <td>
        {reserved ? (
          <Button variant="outline-danger" type="button" onClick={handleButtonClick}>
            Leave Mission
          </Button>
        ) : (
          <Button variant="outline-success" type="button" onClick={handleButtonClick}>
            Join Mission
          </Button>
        )}
      </td>
    </>
  );
};

Mission.propTypes = {
  mission: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
  }).isRequired,
};

export default Mission;
