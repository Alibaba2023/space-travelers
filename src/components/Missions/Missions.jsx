import React, { useEffect } from 'react';
import { Table, Button } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { AllMissions, fetchMissions } from '../../redux/missions/missionSlice';
import '../style/missions.css';
import Mission from './Mission';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector(AllMissions);

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  return (
    <Table striped bordered hover className="missions-table">
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th>Join/Leave</th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={mission.id}>
            <Mission mission={mission} />
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

export default Missions;
