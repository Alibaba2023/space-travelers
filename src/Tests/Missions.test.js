import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import Missions from '../components/Missions/Missions';
import missionsSlice from '../redux/missions/missionSlice';

const store = configureStore({
  reducer: {
    missions: missionsSlice,
  },
});

console.log(store)

describe('Missions', () => {
  test('Component renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}><Missions /></Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Component has spinner and no error', () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    const spinner = screen.getByRole('status');
    expect(spinner).toBeInTheDocument();

    const errorAlert = screen.queryByRole('alert');
    expect(errorAlert).not.toBeInTheDocument();
  });
});
