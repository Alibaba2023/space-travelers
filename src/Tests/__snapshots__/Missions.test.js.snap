// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component renders correctly 1`] = `
<div
  className="spinner-border"
  role="status"
/>
`;
